comp/collector/collector/collectorimpl/collector.go

```diff
@@ -75,3 +104,255 @@ func newNoneCollector() provides {
+
+	if c.state.Load() != started {
+		return emptyID, fmt.Errorf("the collector is not running")
+	}
+
+	if _, found := c.checks[ch.ID()]; found {
+		return emptyID, fmt.Errorf("a check with ID %s is already running", ch.ID())
+	}
+
+	err := c.scheduler.Enter(ch)
+	if err != nil {
+		return emptyID, fmt.Errorf("unable to schedule the check: %s", err)
+	}
+
+	// Track the total number of checks running in order to have an appropriate number of workers
+	c.checkInstances++
+	if ch.Interval() == 0 {
+		// Adding a temporary runner for long running check in case the
+		// number of runners is lower than the number of long running
+		// checks.
+		c.log.Infof("Adding an extra runner for the '%s' long running check", ch)
+		c.runner.AddWorker()
+	} else {
+		c.runner.UpdateNumWorkers(c.checkInstances)
+	}
+
+	c.checks[ch.ID()] = ch
+	c.notify(ch.ID(), collector.CheckRun)
+	return ch.ID(), nil
+}
+
+// StopCheck halts a check and remove the instance
+func (c *collectorImpl) StopCheck(id checkid.ID) error {
+	if !c.started() {
+		return fmt.Errorf("the collector is not running")
+	}
+
+	ch, found := c.get(id)
+	if !found {
+		return fmt.Errorf("cannot find a check with ID %s", id)
+	}
+
+	// unschedule the instance
+	err := c.scheduler.Cancel(id)
+	if err != nil {
+		return fmt.Errorf("an error occurred while canceling the check schedule: %s", err)
+	}
+
+	err = c.runner.StopCheck(id)
+	if err != nil {
+		// still attempt to cancel the check before returning the error
+		_ = c.cancelCheck(ch, c.cancelCheckTimeout)
+		return fmt.Errorf("an error occurred while stopping the check: %s", err)
+	}
+
+	err = c.cancelCheck(ch, c.cancelCheckTimeout)
+	if err != nil {
+		return fmt.Errorf("an error occurred while calling check.Cancel(): %s", err)
+	}
+
+	// remove the check from the stats map
+	expvars.RemoveCheckStats(id)
+
+	// vaporize the check
+	c.delete(id)
+
+	return nil
+}
+
+// cancelCheck calls Cancel on the passed check, with a timeout
+func (c *collectorImpl) cancelCheck(ch check.Check, timeout time.Duration) error {
+	done := make(chan struct{})
+
+	go func() {
+		ch.Cancel()
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		return nil
+	case <-time.After(timeout):
+		return fmt.Errorf("timeout while calling check.Cancel() on check ID %s, timeout: %s", ch.ID(), timeout)
+	}
+}
+
+func (c *collectorImpl) get(id checkid.ID) (check.Check, bool) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+
+	ch, found := c.checks[id]
+	return ch, found
+}
+
+// remove the check from the list
+func (c *collectorImpl) delete(id checkid.ID) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	delete(c.checks, id)
+	c.notify(id, collector.CheckStop)
+}
+
+// lightweight shortcut to see if the collector has started
+func (c *collectorImpl) started() bool {
+	return c.state.Load() == started
+}
+
+// MapOverChecks call the callback with the list of checks locked.
+func (c *collectorImpl) MapOverChecks(cb func([]check.Info)) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+
+	cInfo := []check.Info{}
```

Comment:
What is the reason for copying the slice before passing to the callback? Might be worth leaving a comment to explain.

```suggestion
	cInfo := make([]check.Info{}, 0, len(c.checks))
```

Reply:
I applied the suggestion. 
I assume the reason for the copy is to prevent any possible race conditions against the internal checks list. But i'm not 100% sure.